crates/sui-graphql-rpc/src/types/event.rs

```diff
@@ -322,6 +346,35 @@ impl Target<Cursor> for StoredEvent {
         Cursor::new(EventKey {
             tx: self.tx_sequence_number as u64,
             e: self.event_sequence_number as u64,
+            checkpoint_viewed_at: None,
         })
     }
+
+    fn consistent_cursor(&self, checkpoint_viewed_at: u64) -> Cursor {
+        Cursor::new(EventKey {
+            tx: self.tx_sequence_number as u64,
+            e: self.event_sequence_number as u64,
+            checkpoint_viewed_at: Some(checkpoint_viewed_at),
+        })
+    }
+}
+
+pub(crate) fn validate_cursor_consistency(
+    after: Option<&Cursor>,
+    before: Option<&Cursor>,
+) -> Result<Option<u64>, Error> {
+    match (after, before) {
+        (Some(after_cursor), Some(before_cursor)) => {
+            if after_cursor.checkpoint_viewed_at == before_cursor.checkpoint_viewed_at {
+                Ok(after_cursor.checkpoint_viewed_at)
+            } else {
+                Err(Error::Client(
+                    "Cursors are inconsistent and cannot be used together in the same query."
+                        .to_string(),
+                ))
+            }
+        }
+        (Some(cursor), None) | (None, Some(cursor)) => Ok(cursor.checkpoint_viewed_at),
```

Comment:
Ditto to earlier PR -- shouldn't be an inconsistent case?

Reply:
addressed in https://github.com/MystenLabs/sui/pull/15996